### PR TITLE
feat: Make add piece idempotent

### PR DESCRIPTION
**Pull request by aarshkshah1992**
A very common deal making error we encounter is that Boost/Markets makes a call to `AddPiece` and then restarts.
If it restarts before it's able to persist the fact that it's already called AddPiece, it will call `AddPiece` again for the same piece. In this case, the existing API returns an error as it's not idempotent.

This PR allows the user to call `AddPiece` multiple times for the same piece but ensures that it's handed off to the sealing subsystem only once. 

It also makes the method respect the given context.

### CODE REVIEW

File: extern/storage-sealing/input.go [hunk 1]
```diff
@@ -311,41 +311,45 @@ func (m *Sealing) SectorAddPieceToAny(ctx context.Context, size abi.UnpaddedPiec
...
 	}
 
-	resCh := make(chan struct {
-		sn     abi.SectorNumber
-		offset abi.UnpaddedPieceSize
-		err    error
-	}, 1)
-
-	m.pendingPieces[proposalCID(deal)] = &pendingPiece{
+	doneCh := make(chan struct{}, 1)
```
Comment (dirkmc) [inline extern/storage-sealing/input.go:325]:
```suggestion
	doneCh := make(chan struct{})
```

File: extern/storage-sealing/sealing.go [hunk 2]
```diff
@@ -140,7 +141,16 @@ func (o *openSector) dealFitsInLifetime(dealEnd abi.ChainEpoch, expF func(sn abi
...
 
+type pieceAcceptResp struct {
+	sn     abi.SectorNumber
+	offset abi.UnpaddedPieceSize
+	err    error
+}
+
 type pendingPiece struct {
+	doneCh chan struct{}
+	resp   atomic.Value
```
Comment (dirkmc) [inline extern/storage-sealing/sealing.go:152]: I don't think resp needs to be an atomic value - the doneCh prevents concurrent access

File: extern/storage-sealing/input.go [hunk 3]
```diff
@@ -311,41 +311,45 @@ func (m *Sealing) SectorAddPieceToAny(ctx context.Context, size abi.UnpaddedPiec
...
 
 	m.inputLk.Lock()
-	if _, exist := m.pendingPieces[proposalCID(deal)]; exist {
+	if pp, exist := m.pendingPieces[proposalCID(deal)]; exist {
 		m.inputLk.Unlock()
-		return api.SectorOffset{}, xerrors.Errorf("piece for deal %s already pending", proposalCID(deal))
+		select {
+		case <-pp.doneCh:
+			res := pp.resp.Load().(*pieceAcceptResp)
+			return api.SectorOffset{Sector: res.sn, Offset: res.offset.Padded()}, res.err
```
Comment (dirkmc) [inline extern/storage-sealing/input.go:319]: Is it safe to call `res.offset.Padded()` if res.err is not nil?

File: extern/storage-sealing/input.go (hunk 1 above)
Comment (aarshkshah1992) [reply extern/storage-sealing/input.go:325]: Let's keep this for safety, dosen't hurt because the code that writes to this is called sync by the sealer

File: extern/storage-sealing/sealing.go (hunk 2 above)
Comment (aarshkshah1992) [reply extern/storage-sealing/sealing.go:152]: good point. fixed.

File: extern/storage-sealing/input.go (hunk 3 above)
Comment (aarshkshah1992) [reply extern/storage-sealing/input.go:319]: Yup it is, it's an uint.

File: extern/storage-sealing/input.go [hunk 4]
```diff
@@ -311,41 +311,45 @@ func (m *Sealing) SectorAddPieceToAny(ctx context.Context, size abi.UnpaddedPiec
...
+	if pp, exist := m.pendingPieces[proposalCID(deal)]; exist {
 		m.inputLk.Unlock()
-		return api.SectorOffset{}, xerrors.Errorf("piece for deal %s already pending", proposalCID(deal))
+		select {
+		case <-pp.doneCh:
+			res := pp.resp
+			return api.SectorOffset{Sector: res.sn, Offset: res.offset.Padded()}, res.err
+		case <-ctx.Done():
+			return api.SectorOffset{}, ctx.Err()
+		}
```
Comment (magik6k) [inline extern/storage-sealing/input.go:322]: Could put this into a helper method (e.g. `waitResp`) on `pendingPiece` for readability, and dedupe the same bit of code below.

File: extern/storage-sealing/input.go (hunk 1 above)
Comment (magik6k) [reply extern/storage-sealing/input.go:325]: We never really write to this channel, just close it, which doesn't block.

File: extern/storage-sealing/input.go [hunk 5]
```diff
@@ -311,41 +311,45 @@ func (m *Sealing) SectorAddPieceToAny(ctx context.Context, size abi.UnpaddedPiec
 	}
 
 	m.inputLk.Lock()
-	if _, exist := m.pendingPieces[proposalCID(deal)]; exist {
+	if pp, exist := m.pendingPieces[proposalCID(deal)]; exist {
```
Comment (magik6k) [inline extern/storage-sealing/input.go:314]: So there is one somewhat big issue with this - when a market process calls AddPiece, it passes a reader to the call, and that data will be streamed directly into the sector.

If the market process restarts in the middle, or when we already we have a waiting AddPiece call, the stream will be broken, so here we will be waiting for as AddPiece call which is guaranteed to fail - instead, I think we should:
* Wait like we do now;
* If we see an error, try to add the piece to `m.pendingPieces` again  - we know that our reader was not read from, so all the bytes are in correct places, and the stream can be handed off to workers

Tldr I'd make this if into a for loop trying to add the piece to `m.pendingPieces` as long as:
* After waiting we somehow see no error (which is unlikely if we're here, but can happen with weird timing)
* Context is cancelled
* The previous call errored, so we can now add our piece to `m.pendingPieces` again

**Comment by aarshkshah1992** [comment]
@magik6k  Have addressed your review.

**Comment by jennijuju** [comment]
Please update the pr title as instructed in  the pr template before merging
